### PR TITLE
bpo-37249: add declaration of _PyObject_GetMethod

### PR DESCRIPTION
**Pull request by jdemeyer**
https://bugs.python.org/issue37249

### CODE REVIEW

File: Include/internal/pycore_object.h [hunk 1]
```diff
@@ -10,6 +10,10 @@ extern "C" {
 
 #include "pycore_pystate.h"   /* _PyRuntime */
 
+/* Private API for the LOAD_METHOD opcode. */
```
Comment (methane) [inline Include/internal/pycore_object.h]: I think this comment is not needed.
We have detailed comment in function definition.

File: Include/internal/pycore_object.h [hunk 2]
```diff
@@ -10,6 +10,10 @@ extern "C" {
...
+PyAPI_FUNC(int)
+_PyObject_GetMethod(PyObject *obj, PyObject *name, PyObject **method);
```
Comment (methane) [inline Include/internal/pycore_object.h]: Other API definitions in this file are single line format.

Comment (methane) [reply Include/internal/pycore_object.h]: BTW, I'm not sure we need `PyAPI_FUNC` here.
It is used to expose symbol.  But this API is really private.  This is used only in python executable or python dll.
Even extension modules in stdlib must not use it.

Comment (jdemeyer) [reply Include/internal/pycore_object.h]: > BTW, I'm not sure we need `PyAPI_FUNC` here.

It seems that most functions in the internal API do use `PyAPI_FUNC`. So it seems better to keep it. If you insist, I will remove the `PyAPI_FUNC`.

**Comment by jdemeyer** [comment]
I'm not actually convinced that `_PyObject_GetMethod` should be super-internal. For example, Cython literally copied the code of `_PyObject_GetMethod` from CPython because CPython does not expose it. So maybe we shouldn't actually make it internal.

**Comment by methane** [comment]
I expect further optimization which requires changing this API completely.

For general use, I want more high level API for calling method with vectercall.

So, if expected user is only Cython, please keep copying.

**Comment by jdemeyer** [comment]
> So, if expected user is only Cython, please keep copying.

OK, got it.

**Comment by msullivan** [comment]
> So, if expected user is only Cython, please keep copying.

I'm probably *also* going to copy it for mypyc at some point, but I think that falls in the same bucket.

**Review by msullivan — APPROVED** [review]

**Comment by methane** [comment]
If there are at least two users (mypyc and Cython), I'm OK to move it to private/.

**Comment by jdemeyer** [comment]
> I'm OK to move it to private/.

What do you mean here? There is a directory `cpython/` for the default non-limited API and a directory `internal/` for the API that is only defined when `Py_BUILD_CORE` is defined. Which of the two do you mean?

**Comment by methane** [comment]
I'm sorry, there is no `private/` directory.  As @vstinner said:

```
Python 3.8 now has a better separation between "private" and "internal" 
APIs:

* private are "_Py" symbols which are exported in the python DLL: they 
should not be used, but can be used techically

* internal are symbols defined in internal header files 
(Include/internal/). Some symbols use PyAPI_FUNC()/PyAPI_DATA(), some 
only use "extern".
```

So I meant put this API in `Include/cpython/object.h`, near to `_PyObject_GetAttrId`.

I will add new really internal API to implement per-opcode cache for LOAD_METHOD,
instead of changing existing `_PyObject_GetMethod`.

**Comment by jdemeyer** [comment]
> So I meant put this API in `Include/cpython/object.h`, near to `_PyObject_GetAttrId`.

OK, done.

**Pull request merged**